### ISSUE TITLE
A problem of EAP-TTLS

### ISSUE THREAD

**Issue by Du-lumos** [bcgit/bc-java#1359]
Recently,I need to write EAP-TTLS's server, my company used BounyCastle to implement PEAP'server、TLS's server. During the development of TTLS, i find that i back a accept packet, but ac show the "authentication fail". I checked packet 、username、password，but nothing exception.
I doubt material key has some of problems. 
E is there any special difference between TTLS and PEAP in the handshake phase when using Bouncy Castle, except for generating the label of the key material and the client_random and server_random's order.,
This my implement: in notifyHandshakeComplete()

`public static MPPEKeys deriveMPPEKeysByTlsContext(TlsContext tlsContext , AccessRequest accessRequest) {
        MPPEKeys derivedMPPEKeys = new MPPEKeys();
        byte[] keyMaterial = null;
        if (TLS_VER_10.equals(tlsContext.getClientVersion().getName())) {
            keyMaterial = exportKeyingMaterial(ExporterLabel.ttls_keying_material, null, 64, tlsContext.getSecurityParametersConnection() , null);
        } else {
            keyMaterial = tlsContext.exportKeyingMaterial(ExporterLabel.ttls_keying_material, null, 128);
        }
        byte[] recvKey = new byte[32];
        byte[] sendKey = new byte[32];
        System.arraycopy(keyMaterial, 0, recvKey, 0, 32);
        System.arraycopy(keyMaterial, 32, sendKey, 0, 32);
        derivedMPPEKeys.setRecvKey(recvKey);
        derivedMPPEKeys.setSendKey(sendKey);
        return derivedMPPEKeys;
    }`

**Comment by peterdettman** [comment]
Not sure on your actual question, but if you are trying to use different code when TLS 1.0 is negotiatied, then:

`if (TLS_VER_10.equals(tlsContext.getClientVersion().getName()))`

should probably be:

`if (!TlsUtils.isTLSv11(tlsContext))`

**Comment by peterdettman** [comment]
I am happy you got this working, but it is a very strange fix. The calculation should not give a different result from doing it all in `notifyHandshakeComplete`.

**Comment by Du-lumos** [comment]
> 

Yeah,it's strange fix. I doubt that TTLS is called 'tunnel tls', and them client_random and server_random should be draw in the tunnel handshake process, rather than the tunnel handshake complele.
A significate difference in the BouncyCastle, In handshake process, we want to get the server/client random from 'getTlsContext().getSecurityParametersHandshake().getClientRandom();',but after the handshake complete,we want to get the server/client random from tlsContext.getSecurityParametersConnection().getClientRandom().

**Comment by peterdettman** [comment]
If you check the values of clientRandom/serverRandom you should see that the value is the same for both ways.

**Comment by Du-lumos** [comment]
> If you check the values of clientRandom/serverRandom you should see that the value is the same for both ways.

Yeah, it's different

**Comment by winfriedgerlach** [comment]
@peterdettman @Du-lumos can this ticket be closed?
